src/Riok.Mapperly/Descriptors/MappingBuilders/SpanMappingBuilder.cs

```diff
@@ -0,0 +1,212 @@
+                ctx.Source,
+                ctx.Types.Get(typeof(List<>)),
+                ctx.Types
+            );
+            return new ForEachAddEnumerableMapping(
+                ctx.Source,
+                ctx.Types.Get(typeof(List<>)).Construct(elementMapping.TargetType),
+                elementMapping,
+                null,
+                AddMethodName,
+                ensureCapacityStatement
+            );
+        }
+
+        if (
+            !ctx.ObjectFactories.TryFindObjectFactory(ctx.Source, ctx.Target, out var objectFactory)
+            && !ctx.Target.HasAccessibleParameterlessConstructor()
+        )
+        {
+            return MapSpanArrayToEnumerableMethod(ctx);
+        }
+
+        if (target.Type is CollectionType.Stack)
+            return CreateForEach(nameof(Stack<object>.Push), objectFactory);
+
+        if (target.Type is CollectionType.Queue)
+            return CreateForEach(nameof(Queue<object>.Enqueue), objectFactory);
+
+        // create a foreach loop with add calls if source is not an array
+        // and ICollection.Add(T): void is implemented and not explicit
+        // ensures add is not called and immutable types
+        if (
+            target.Type is not CollectionType.Array
+            && ctx.Target.HasImplicitGenericImplementation(ctx.Types.Get(typeof(ICollection<>)), AddMethodName)
+        )
+            return CreateForEach(AddMethodName, objectFactory);
+
+        return MapSpanArrayToEnumerableMethod(ctx);
+
+        ExistingTargetMappingMethodWrapper CreateForEach(string methodName, ObjectFactory? factory)
+        {
+            var ensureCapacityStatement = EnsureCapacityBuilder.TryBuildEnsureCapacity(ctx);
+            return new ForEachAddEnumerableMapping(ctx.Source, ctx.Target, elementMapping, factory, methodName, ensureCapacityStatement);
+        }
+    }
+
+    private static TypeMapping BuildToArrayOrMap(MappingBuilderContext ctx, ITypeMapping elementMapping)
+    {
+        if (!elementMapping.IsSynthetic)
+            return new ArrayForMapping(ctx.Source, ctx.Target, elementMapping, elementMapping.TargetType);
+
+        return new SourceObjectMethodMapping(ctx.Source, ctx.Target, ToArrayMethodName);
+    }
+
+    private static TypeMapping? BuildEnumerableToSpan(MappingBuilderContext ctx, ITypeMapping elementMapping)
+    {
+        var typedArray = ctx.Types.GetArrayType(elementMapping.TargetType);
+        if (ctx.FindOrBuildMapping(ctx.Source, typedArray.NonNullable()) is not { } arrayMapping)
+            return null;
+
+        return new CastMapping(ctx.Source, ctx.Target, arrayMapping);
+    }
+
+    private static TypeMapping? MapSpanArrayToEnumerableMethod(MappingBuilderContext ctx)
+    {
+        var enumeratedType = ctx.CollectionInfos!.Source.EnumeratedType;
+        var typedArray = ctx.Types.GetArrayType(enumeratedType);
+        if (ctx.FindOrBuildMapping(typedArray.NonNullable(), ctx.Target) is not { } arrayMapping)
```

Comment:
Is the `NonNullable()` really correct here? `ToArray()` can never return `null`, can it? Probably `GetArrayType` should always return non-nullable type (while still copying the element nullability). Same on L193.